### PR TITLE
Include underlying type in enum decl syntax when generating it from a symbol

### PR DESCRIPTION
**Pull request by safern**
Currently if an enum symbol contains an underlying type the generated syntax doesn't contain that information.

i.e:
```cs
public enum foo : byte { }
```
we would get back:
```cs
public enum foo { }
```

This fixes that but still doesn't return it if the underlying type is Int32.

### CODE REVIEW

File: src/Workspaces/VisualBasicTest/CodeGeneration/SyntaxGeneratorTests.vb [hunk 1]
```diff
@@ -2275,12 +2275,21 @@ End Interface")
...
     Utc = 1
     Local = 2
+End Enum")
+        End Sub
+
+        <Fact>
+        Public Sub TestEnumWithUnderlyingTypeFromSymbol()
+            VerifySyntax(Of EnumBlockSyntax)(Generator.Declaration(_emptyCompilation.GetTypeByMetadataName("System.Security.SecurityRuleSet")),
+"Public Enum SecurityRuleSet As System.Byte
+    None = CByte(0)
```
Comment (safern) [inline src/Workspaces/VisualBasicTest/CodeGeneration/SyntaxGeneratorTests.vb:2290]: For some reason in VB it is returning a redundant cast.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 2]
```diff
@@ -820,11 +820,18 @@ internal override SyntaxNode AsInterfaceMember(SyntaxNode m)
             DeclarationModifiers modifiers,
             IEnumerable<SyntaxNode> members)
         {
+            return EnumDeclaration(name, null, accessibility, modifiers, members);
+        }
+
+        internal override SyntaxNode EnumDeclaration(string name, SyntaxNode underlyingType, Accessibility accessibility = Accessibility.NotApplicable, DeclarationModifiers modifiers = default, IEnumerable<SyntaxNode> members = null)
+        {
+            List<BaseTypeSyntax> underlyingTypeList = underlyingType != null ? new() { SyntaxFactory.SimpleBaseType((TypeSyntax)underlyingType) } : null;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:828]: there can only be at most one base type, so i don't see the need for the list.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 3]
```diff
@@ -820,11 +820,18 @@ internal override SyntaxNode AsInterfaceMember(SyntaxNode m)
             DeclarationModifiers modifiers,
             IEnumerable<SyntaxNode> members)
         {
+            return EnumDeclaration(name, null, accessibility, modifiers, members);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:823]:
```suggestion
            return EnumDeclaration(name, underlyingType: null, accessibility, modifiers, members);
```

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 4]
```diff
@@ -820,11 +820,18 @@ internal override SyntaxNode AsInterfaceMember(SyntaxNode m)
...
+        internal override SyntaxNode EnumDeclaration(string name, SyntaxNode underlyingType, Accessibility accessibility = Accessibility.NotApplicable, DeclarationModifiers modifiers = default, IEnumerable<SyntaxNode> members = null)
+        {
+            List<BaseTypeSyntax> underlyingTypeList = underlyingType != null ? new() { SyntaxFactory.SimpleBaseType((TypeSyntax)underlyingType) } : null;
+
             return SyntaxFactory.EnumDeclaration(
                 default,
                 AsModifierList(accessibility, modifiers, SyntaxKind.EnumDeclaration),
                 name.ToIdentifierToken(),
-                null,
+                underlyingTypeList != null ? SyntaxFactory.BaseList(SyntaxFactory.SeparatedList(underlyingTypeList)) : null,
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:834]: `                underlyingType != null ? SyntaxFactory.BaseList(SyntaxFactory.SingletonSeparatedList(...etc....underlyingType)) : null`

File: src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs [hunk 5]
```diff
@@ -1823,6 +1823,20 @@ public void TestEnumDeclarationFromSymbol()
 }");
         }
 
+        [Fact]
+        public void TestEnumWithUnderlyingTypeFromSymbol()
+        {
+            VerifySyntax<EnumDeclarationSyntax>(
+                    Generator.Declaration(
+                        _emptyCompilation.GetTypeByMetadataName("System.Security.SecurityRuleSet")),
+@"public enum SecurityRuleSet : global::System.Byte
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs:1832]: nope: this won't parse.  the base type must be an actual primitive like `byte`.  we'll likely have to special case this in the generator. :)

File: src/Workspaces/VisualBasicTest/CodeGeneration/SyntaxGeneratorTests.vb (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/VisualBasicTest/CodeGeneration/SyntaxGeneratorTests.vb:2290]: that's ok.  not an issue for you to deal with. :)

File: src/Workspaces/VisualBasic/Portable/CodeGeneration/VisualBasicSyntaxGenerator.vb [hunk 6]
```diff
@@ -1445,12 +1445,26 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeGeneration
...
+        Friend Overrides Function EnumDeclaration(name As String,
+                                                  underlyingType As SyntaxNode,
+                                                  Optional accessibility As Accessibility = Accessibility.NotApplicable,
+                                                  Optional modifiers As DeclarationModifiers = Nothing,
+                                                  Optional members As IEnumerable(Of SyntaxNode) = Nothing) As SyntaxNode
+            Dim underlyingTypeClause As AsClauseSyntax = Nothing
+
+            If underlyingType IsNot Nothing Then
+                underlyingTypeClause = SyntaxFactory.SimpleAsClause(DirectCast(underlyingType, TypeSyntax))
+            End If
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/CodeGeneration/VisualBasicSyntaxGenerator.vb:1460]:
```suggestion
            Dim underlyingTypeClause = If(underlyingType is nothing, nothing, SyntaxFactory.SimpleAsClause(DirectCast(underlyingType, TypeSyntax)))
```

File: src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs (hunk 5 above)
Comment (safern) [reply src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs:1832]: Ok, I'll look into it. :)

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 2 above)
Comment (safern) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:828]: Ah, I thought I needed the list to create the separated list with the syntax factory, but you mentioned how to do it in another comment. Thanks :)

File: src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs (hunk 5 above)
Comment (safern) [reply src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs:1832]: Interesting, when I tried it in sharplab it did parse it correctly, but makes sense to return it like an actual primitive.
https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEDMACOA7ArmDAMQHtiMQMBzAG2OAmpBBQEYA6AIQE94MBvDAF8gA=

File: src/Workspaces/VisualBasicTest/CodeGeneration/SyntaxGeneratorTests.vb (hunk 1 above)
Comment (safern) [reply src/Workspaces/VisualBasicTest/CodeGeneration/SyntaxGeneratorTests.vb:2290]: Sounds good :)

File: src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs (hunk 5 above)
Comment (safern) [reply src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs]: Fixed.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**